File: scripts/render/policy.py
"""/policy-match/ — step-by-step policy questionnaire (DESIGN.md §7.5).

SSR renders the full question bank (every step and choice) plus the
methodology, so crawlers and no-JS readers see everything. pages/policy.js
turns it into a one-question-at-a-time flow and computes the results in the
browser from a slim inline scoring blob; evidence excerpts are fetched from
/api/policy-match.json. Choices are shared through ``#c=<choice ids>``.
"""

from __future__ import annotations

import json
import shutil
from typing import Any

from . import shell as S

ROUTE = "/policy-match/"
OUT_DIR = S.SITE_ROOT / "policy-match"

LEDE = "依序回答四題，選出你在意的市政方向，再看各城市候選人的主動政策發文和這些方向有多少重合。"
DESCRIPTION = ("先選市政優先方向，再看六都市長候選人主動發起的政策倡議貼文與你的選擇有多少議題重合；"
               "每項結果都附原文證據，資料不足者不推定立場。")

# Shown before JS runs: marks the document as enhanced (hides later steps via
# CSS, no flash) and, for shared links, hides the quiz until results render.
BOOT_SCRIPT = (
    "<script>(function(){var d=document.documentElement;d.classList.add('pm-js');"
    "if(/(^#|&)c=[^&]/.test(location.hash))d.classList.add('pm-hash');})();</script>"
)


def _json_script(element_id: str, payload: Any) -> str:
    text = json.dumps(payload, ensure_ascii=False, separators=(",", ":")).replace("</", "<\\/")
    return f'<script type="application/json" id="{element_id}">{text}</script>'


def _scoring_blob(data: Any, pm: dict[str, Any]) -> dict[str, Any]:
    """Everything policy.js needs to rank candidates without the 300 KB API file."""
    rows = {r.get("candidateId"): r for r in pm.get("candidates") or []}
    candidates = []
    for cand in data.candidates:  # roster only; missing rows → eligible 0 (資料不足)
        row = rows.get(cand["id"]) or {}
        candidates.append({
            "id": cand["id"],
            "name": cand.get("name") or row.get("candidateName") or cand["id"],
            "city": cand.get("city") or row.get("city") or "",
            "party": cand.get("party") or "",
            "avatarUrl": S.asset_abs(cand.get("avatarUrl")),
            "eligible": int(row.get("eligiblePostCount") or 0),
            "weights": {k: v for k, v in (row.get("topicWeights") or {}).items() if v},
        })
    questions = [
        {"id": q.get("id"), "max": int(q.get("maxChoices") or 1),
         "choices": [{"id": c.get("id"), "label": c.get("label"), "topics": c.get("topics") or []}
                     for c in q.get("choices") or []]}
        for q in pm.get("questions") or []
    ]
    return {
        "questions": questions,
        "candidates": candidates,
        "cities": [{"id": c["id"], "label": S.city_label(c["id"], short=False)} for c in data.cities],
        "topicSlugs": S.TOPIC_SLUGS,
    }


def _choice(choice: dict[str, Any]) -> str:
    topics = "、".join(choice.get("topics") or [])
    return (
        f'<button type="button" class="policy-choice" data-choice="{S.esc(choice.get("id"))}" aria-pressed="false">'
        '<span class="policy-check" aria-hidden="true"></span>'
        f'<span class="policy-choice-text"><span class="policy-choice-label">{S.esc(choice.get("label"))}</span>'
        f'<span class="policy-choice-topics">對應議題：{S.esc(topics)}</span></span></button>'
    )


def _step(index: int, total: int, question: dict[str, Any]) -> str:
    qid = S.esc(question.get("id"))
    max_choices = int(question.get("maxChoices") or 1)
    rule = f"可複選，最多 {max_choices} 項" if max_choices > 1 else "單選"
    choices = "".join(_choice(c) for c in question.get("choices") or [])
    current = " is-current" if index == 0 else ""
    return (
        f'<li class="policy-step{current}" data-step="{index}" data-question="{qid}" data-max="{max_choices}">'
        f'<p class="policy-step-kicker">第 {index + 1} 題<span aria-hidden="true"> · </span>'
        f'<span class="policy-rule">{rule}</span></p>'
        f'<h2 class="policy-prompt" id="pq-{qid}" tabindex="-1">{S.esc(question.get("prompt"))}</h2>'
        f'<div class="policy-choices" role="group" aria-labelledby="pq-{qid}">{choices}</div>'
        f'<p class="policy-hint" data-hint aria-live="polite">{rule}；沒有想法也可以略過。</p>'
        "</li>"
    )


def _quiz(questions: list[dict[str, Any]]) -> str:
    total = len(questions)
    steps = "".join(_step(i, total, q) for i, q in enumerate(questions))
    first = 1 / total if total else 0
    return (
        '<section class="policy-quiz card" id="policy-quiz" aria-label="問卷">'
        '<div class="policy-progress">'
        f'<p class="policy-progress-label"><span data-progress-text>第 1 / {total} 題</span>'
        '<span class="policy-picked-count" data-picked-count></span></p>'
        f'<div class="meter" role="progressbar" aria-label="作答進度" aria-valuemin="1" aria-valuemax="{total}" '
        f'aria-valuenow="1" data-progress><span class="meter-fill" style="--v:{first:.4f}"></span></div>'
        "</div>"
        f'<ol class="policy-steps">{steps}</ol>'
        '<div class="policy-nav">'
        f'<button type="button" class="btn" data-act="prev">{S.icon("chevron", "icon policy-prev-icon")}上一題</button>'
        '<button type="button" class="btn btn-primary" data-act="next"><span data-label>下一題</span>'
        f'{S.icon("chevron", "icon policy-next-icon")}</button>'
        '<button type="button" class="btn btn-primary" data-act="result" hidden>看結果</button>'
        "</div>"
        "</section>"
    )


def _results_shell() -> str:
    return (
        '<section class="policy-results" id="policy-results" aria-labelledby="policy-results-title" hidden>'
        '<div class="policy-results-head card">'
        '<div class="policy-results-title-row">'
        '<h2 id="policy-results-title" tabindex="-1">你選的方向</h2>'
        f'<button type="button" class="btn btn-sm" data-act="share">{S.icon("share")}分享結果</button>'
        "</div>"
        '<div class="policy-picked" data-picked></div>'
        '<p class="policy-results-note">重合度只計算候選人<strong>主動發文</strong>中含政策倡議的貼文'
        "（不含回應他方觀點的貼文），比較議題分布，不代表候選人對政策的立場或支持程度；"
        "貼文較少或未提及某議題，不等於反對。</p>"
        "</div>"
        '<div class="policy-cities" data-results></div>'
        '<div class="policy-again">'
        '<button type="button" class="btn" data-act="edit">修改選擇</button>'
        '<button type="button" class="btn btn-primary" data-act="restart">重新選擇</button>'
        "</div>"
        "</section>"
    )


def _methodology(pm: dict[str, Any], data: Any) -> str:
    method = pm.get("methodology") or ""
    maxes = sorted({int(q.get("maxChoices") or 1) for q in pm.get("questions") or []})
    rule = f"每題最多選 {maxes[0]} 項，可略過" if len(maxes) == 1 else "每題有選擇上限，可略過"
    stamp = S.fmt_time_tpe(data.generated_at, with_year=True)
    return (
        '<section class="policy-method" aria-labelledby="policy-method-title">'
        '<h2 id="policy-method-title">計算方式</h2>'
        f'<p>{S.esc(method)}</p>'
        '<ul class="policy-method-list">'
        "<li>只納入候選人<strong>主動發文</strong>、且 AI 判定含政策倡議的貼文；回應他方觀點的貼文與「生活」類不計入。</li>"
        f"<li>你的選擇會展開成對應議題（{rule}），各議題權重相同；"
        "與每位候選人的議題分布計算餘弦相似度，顯示為「議題重合」百分比。</li>"
        "<li>候選人依城市分組，同城市內依重合度排序，同分時依候選人代碼排序；不以政黨排序或配色。</li>"
        "<li>每個議題最多列出 3 則證據貼文摘錄，一律連回原文。沒有可計算貼文的候選人標示「資料不足，不推定立場」。</li>"
        "<li>選擇只存在網址的 <code>#</code> 片段，不會送到伺服器；本頁不使用 cookie 或追蹤。</li>"
        "</ul>"
        f'<p class="snapshot">資料快照：{S.esc(stamp)}（GMT+8）・'
        '<a href="/api/policy-match.json">policy-match.json</a>・<a href="/about/">關於分類方式</a></p>'
        "</section>"
    )


def render(data: Any) -> None:
    shutil.rmtree(OUT_DIR, ignore_errors=True)
    pm = data.policy_match or {}
    title = pm.get("title") or "你的市政優先順序"
    questions = pm.get("questions") or []

    if questions:
        main = _quiz(questions) + _results_shell()
    else:
        main = S.render_empty("尚無資料", "政策題庫尚未產生，請稍後再來。")

    body = (
        BOOT_SCRIPT
        + S.page_head(title, S.esc(LEDE), eyebrow="議題選擇器")
        + '<noscript><p class="notice policy-noscript">逐題作答與結果計算需要啟用 JavaScript；'
          "以下仍列出完整題目與計算方式。</p></noscript>"
        + '<div class="policy-app" id="policy-app">' + main + "</div>"
        + _methodology(pm, data)
        + _json_script("policy-data", _scoring_blob(data, pm))
    )
    jsonld = {
        "@context": "https://schema.org", "@type": "WebPage", "name": f"{title}｜{S.SITE_NAME}",
        "url": S.canonical(ROUTE), "description": DESCRIPTION, "inLanguage": "zh-Hant",
    }
    html = S.layout("policy", title, DESCRIPTION, ROUTE, body,
                    extra_css=["pages/policy.css"], extra_js=["pages/policy.js"], jsonld=jsonld)
    S.write_page(ROUTE, html)

File: site/assets/pages/policy.js
/* /policy-match/ — step-by-step questionnaire + results (DESIGN.md §7.5).
   SSR ships every question; this script shows one step at a time, scores
   candidates from the inline #policy-data blob and fetches evidence excerpts
   from /api/policy-match.json. State lives only in the URL hash (#c=ids). */
(function () {
  "use strict";
  var MO = window.MO;
  var root = document.documentElement;
  var app = document.getElementById("policy-app");
  var dataEl = document.getElementById("policy-data");
  if (!MO || !app || !dataEl) { root.classList.remove("pm-js", "pm-hash"); return; }

  var D;
  try { D = JSON.parse(dataEl.textContent); } catch (e) { root.classList.remove("pm-js", "pm-hash"); return; }
  var esc = MO.esc, qs = MO.qs, qsa = MO.qsa;

  var quiz = qs("#policy-quiz", app);
  var results = qs("#policy-results", app);
  var steps = qsa(".policy-step", quiz);
  var total = steps.length;
  if (!quiz || !results || !total || !Array.isArray(D.questions) || !D.questions.length) {
    root.classList.remove("pm-js", "pm-hash"); return;
  }
  var btnPrev = qs('[data-act="prev"]', quiz);
  var btnNext = qs('[data-act="next"]', quiz);
  var btnResult = qs('[data-act="result"]', quiz);
  var progressText = qs("[data-progress-text]", quiz);
  var progressBar = qs("[data-progress]", quiz);
  var pickedCount = qs("[data-picked-count]", quiz);

  var choiceById = {};
  var questionOf = {};
  D.questions.forEach(function (q) {
    q.choices.forEach(function (c) { choiceById[c.id] = c; questionOf[c.id] = q; });
  });

  /* selection: question id → ordered array of choice ids */
  var sel = {};
  var current = 0;

  function selectedIds() {
    var out = [];
    D.questions.forEach(function (q) { (sel[q.id] || []).forEach(function (id) { out.push(id); }); });
    return out;
  }

  /* ---------------- quiz ---------------- */
  function syncStep(i) {
    var step = steps[i];
    var q = D.questions[i];
    var picked = sel[q.id] || [];
    var full = picked.length >= q.max;
    qsa(".policy-choice", step).forEach(function (b) {
      var on = picked.indexOf(b.getAttribute("data-choice")) !== -1;
      b.setAttribute("aria-pressed", on ? "true" : "false");
      if (!on && full && q.max > 1) b.setAttribute("aria-disabled", "true");
      else b.removeAttribute("aria-disabled");
    });
  }

  function hint(i, text) {
    var el = qs("[data-hint]", steps[i]);
    var q = D.questions[i];
    if (text) { el.textContent = text; el.classList.add("is-alert"); return; }
    el.classList.remove("is-alert");
    var n = (sel[q.id] || []).length;
    var rule = q.max > 1 ? "可複選，最多 " + q.max + " 項" : "單選";
    el.textContent = n ? "已選 " + n + " / " + q.max + " 項" + (n >= q.max && q.max > 1 ? "；要換選項請先取消一項。" : "。")
                       : rule + "；沒有想法也可以略過。";
  }

  function show(i, focus) {
    current = Math.max(0, Math.min(total - 1, i));
    steps.forEach(function (s, k) { s.classList.toggle("is-current", k === current); });
    var last = current === total - 1;
    var n = selectedIds().length;
    btnPrev.disabled = current === 0;
    btnNext.hidden = last;
    btnResult.hidden = !last;
    btnResult.disabled = n === 0;
    var q = D.questions[current];
    var has = (sel[q.id] || []).length > 0;
    qs("[data-label]", btnNext).textContent = has ? "下一題" : "略過這題";
    btnNext.classList.toggle("btn-primary", has);
    progressText.textContent = "第 " + (current + 1) + " / " + total + " 題";
    progressBar.setAttribute("aria-valuenow", String(current + 1));
    qs(".meter-fill", progressBar).style.setProperty("--v", ((current + 1) / total).toFixed(4));
    pickedCount.textContent = n ? "已選 " + n + " 項" : "";
    syncStep(current);
    hint(current, last && n === 0 ? "還沒有選任何方向；至少選一項才能看結果。" : "");
    if (focus) {
      var h = qs(".policy-prompt", steps[current]);
      if (h) h.focus({ preventScroll: true });
      quiz.scrollIntoView({ block: "nearest" });
    }
  }

  function toggle(btn) {
    var id = btn.getAttribute("data-choice");
    var q = questionOf[id];
    if (!q) return;
    var arr = (sel[q.id] = sel[q.id] || []);
    var at = arr.indexOf(id);
    if (at !== -1) arr.splice(at, 1);
    else if (q.max === 1) arr.splice(0, arr.length, id);
    else if (arr.length >= q.max) {
      syncStep(current);
      hint(current, "這題最多選 " + q.max + " 項，請先取消已選的一項。");
      return;
    } else arr.push(id);
    show(current, false);
  }

  quiz.addEventListener("click", function (ev) {
    var choice = ev.target.closest(".policy-choice");
    if (choice) { toggle(choice); return; }
    var act = ev.target.closest("[data-act]");
    if (!act || act.disabled) return;
    var a = act.getAttribute("data-act");
    if (a === "prev") show(current - 1, true);
    else if (a === "next") show(current + 1, true);
    else if (a === "result") showResults(true);
  });
  /* ← / → move between choices inside a step (buttons stay in Tab order too) */
  quiz.addEventListener("keydown", function (ev) {
    if (ev.key !== "ArrowRight" && ev.key !== "ArrowLeft" && ev.key !== "ArrowDown" && ev.key !== "ArrowUp") return;
    var choice = ev.target.closest && ev.target.closest(".policy-choice");
    if (!choice) return;
    var list = qsa(".policy-choice", steps[current]);
    var k = list.indexOf(choice) + (ev.key === "ArrowRight" || ev.key === "ArrowDown" ? 1 : -1);
    if (k < 0 || k >= list.length) return;
    ev.preventDefault();
    list[k].focus();
  });

  /* ---------------- hash ---------------- */
  function readHash() {
    var m = /(?:^#|&)c=([^&]*)/.exec(location.hash);
    if (!m) return null;
    var ids;
    try { ids = decodeURIComponent(m[1]).split(","); } catch (e) { return null; }
    var next = {};
    ids.forEach(function (id) {
      id = id.trim();
      var q = questionOf[id];
      if (!q) return;
      var arr = (next[q.id] = next[q.id] || []);
      if (arr.indexOf(id) === -1 && arr.length < q.max) arr.push(id);
    });
    return next;
  }
  function writeHash(ids) {
    var url = location.pathname + location.search + (ids.length ? "#c=" + ids.join(",") : "");
    try { history.replaceState(null, "", url); } catch (e) { location.hash = ids.length ? "c=" + ids.join(",") : ""; }
  }

  /* ---------------- scoring ---------------- */
  function unionTopics(ids) {
    var seen = {}, out = [];
    ids.forEach(function (id) {
      (choiceById[id].topics || []).forEach(function (t) { if (!seen[t]) { seen[t] = 1; out.push(t); } });
    });
    return out;
  }
  /* cosine between the user's uniform topic vector and the candidate's weights */
  function score(c, topics) {
    if (!c.eligible || !topics.length) return null;
    var w = c.weights || {};
    var norm = Math.sqrt(Object.keys(w).reduce(function (s, k) { return s + w[k] * w[k]; }, 0));
    if (!norm) return 0;
    var dot = topics.reduce(function (s, t) { return s + (w[t] || 0); }, 0);
    return dot / (Math.sqrt(topics.length) * norm);
  }

  /* ---------------- evidence (lazy) ---------------- */
  var evidence = null;       // candidateId → {topic: [items]}
  var evidenceFailed = false;
  function loadEvidence() {
    return MO.fetchJSON("/api/policy-match.json").then(function (pm) {
      evidence = {};
      (pm.candidates || []).forEach(function (r) { evidence[r.candidateId] = r.evidence || {}; });
    }, function () { evidenceFailed = true; });
  }
  var evidenceReady = loadEvidence();

  function platformOf(item) {
    var p = String(item.postId || "").split(":")[0];
    return MO.PLATFORM_LABELS && MO.PLATFORM_LABELS[p] ? p : "website";
  }
  function evidenceHTML(cid, topic) {
    if (evidenceFailed) return '<p class="policy-ev-empty">證據貼文暫時無法取得，請重新整理頁面。</p>';
    if (!evidence) return "";
    var items = ((evidence[cid] || {})[topic] || []).slice(0, 3);
    if (!items.length) return '<p class="policy-ev-empty">這個議題沒有可列出的證據貼文。</p>';
    return '<ul class="policy-ev-list">' + items.map(function (it) {
      var plat = platformOf(it);
      var label = (MO.PLATFORM_LABELS && MO.PLATFORM_LABELS[plat]) || "原文";
      var text = String(it.text || "").replace(/\s+/g, " ").trim();
      var more = text.length >= 180 ? "…" : "";
      var href = MO.safeUrl(it.url, true);
      return '<li class="policy-ev"><p class="policy-ev-text">' + esc(text) + more + "</p>" +
        (href ? '<a class="feed-action policy-ev-link" href="' + esc(href) + '" target="_blank" rel="noopener" ' +
        'aria-label="在 ' + esc(label) + ' 開啟原文：' + esc(text.slice(0, 24)) + '">' +
        MO.icon(plat) + "<span>" + esc(label) + " 原文</span>" + MO.icon("external") + "</a>" : "") + "</li>";
    }).join("") + "</ul>";
  }

  /* ---------------- results ---------------- */
  var lastTopics = [];

  function pct(v) { return Math.round(v * 100); }

  function candidateHTML(c, topics) {
    var href = "/" + c.city + "/" + c.id + "/";
    var s = c.score;
    var head = '<header class="policy-cand-head">' +
      MO.avatarHTML(c, "md", href, "policy-cand-avatar") +
      '<div class="policy-cand-id"><a class="policy-cand-name" href="' + esc(href) + '">' + esc(c.name) + "</a>" +
      '<span class="chip-soft chip-party" data-party="' + MO.partySlug(c.party) + '"><span>' + esc(c.party || "無黨籍") + "</span></span></div>";
    if (s === null) {
      return '<article class="policy-cand card is-insufficient" data-candidate="' + esc(c.id) + '">' + head +
        '<p class="policy-score policy-score-na">資料不足</p></header>' +
        '<p class="policy-insufficient">資料不足，不推定立場。目前沒有可計算的主動政策倡議貼文。</p></article>';
    }
    head += '<p class="policy-score"><span class="policy-score-num">' + pct(s) + '<span class="policy-score-unit">%</span></span>' +
      '<span class="policy-score-label">議題重合</span></p></header>';
    var meter = '<div class="meter meter-lg" role="img" aria-label="議題重合 ' + pct(s) + '%"><span class="meter-fill" style="--v:' +
      Math.max(0, Math.min(1, s)).toFixed(4) + '"></span></div>' +
      '<p class="policy-basis">依 ' + c.eligible + " 則主動政策倡議貼文計算</p>";
    var w = c.weights || {};
    var matched = topics.filter(function (t) { return w[t]; }).sort(function (a, b) { return (w[b] - w[a]) || (a < b ? -1 : 1); });
    var missing = topics.filter(function (t) { return !w[t]; });
    var body = "";
    if (matched.length) {
      body += '<div class="policy-topics" role="group" aria-label="' + esc(c.name) + ' 的相符議題（選擇以查看證據）">' +
        matched.map(function (t, k) {
          return '<button type="button" class="chip policy-topic" data-topic="' + esc(t) + '" aria-pressed="' + (k === 0 ? "true" : "false") + '">' +
            "<span>" + esc(t) + '</span><span class="chip-count">' + pct(w[t]) + "%</span></button>";
        }).join("") + "</div>" +
        '<p class="policy-topics-hint">百分比＝該議題佔其政策貼文的比例</p>' +
        '<div class="policy-ev-wrap" data-ev>' + evidenceHTML(c.id, matched[0]) + "</div>";
    } else {
      body += '<p class="policy-ev-empty">所選方向的議題，在其主動政策貼文中都沒有出現。這不代表反對。</p>';
    }
    if (missing.length && matched.length) {
      body += '<p class="policy-missing">未見相關貼文：' + missing.map(esc).join("、") + "</p>";
    }
    return '<article class="policy-cand card" data-candidate="' + esc(c.id) + '">' + head + meter + body + "</article>";
  }

  function renderResults(ids) {
    var topics = unionTopics(ids);
    lastTopics = topics;
    qs("[data-picked]", results).innerHTML =
      '<ul class="policy-picked-list">' + ids.map(function (id) {
        return '<li class="chip-soft policy-picked-chip"><span>' + esc(choiceById[id].label) + "</span></li>";
      }).join("") + "</ul>" +
      '<p class="policy-picked-topics">對應議題：' + topics.map(function (t) {
        return '<a href="/spectrum/' + esc((D.topicSlugs || {})[t] || "") + '/">' + esc(t) + "</a>";
      }).join("、") + "</p>";

    var byCity = {};
    D.candidates.forEach(function (c) {
      var row = { id: c.id, name: c.name, city: c.city, party: c.party, avatarUrl: c.avatarUrl,
                  eligible: c.eligible, weights: c.weights, score: score(c, topics) };
      (byCity[c.city] = byCity[c.city] || []).push(row);
    });
    var html = D.cities.map(function (city) {
      var rows = byCity[city.id] || [];
      if (!rows.length) return "";
      rows.sort(function (a, b) {
        var sa = a.score === null ? -1 : a.score, sb = b.score === null ? -1 : b.score;
        return (sb - sa) || (a.id < b.id ? -1 : a.id > b.id ? 1 : 0);
      });
      return '<section class="policy-city" data-city="' + esc(city.id) + '" aria-labelledby="pc-' + esc(city.id) + '">' +
        '<h3 class="policy-city-title" id="pc-' + esc(city.id) + '"><span class="policy-city-dot" aria-hidden="true"></span>' +
        esc(city.label) + '<span class="policy-city-count">' + rows.length + " 位候選人</span></h3>" +
        '<div class="policy-cands">' + rows.map(function (r) { return candidateHTML(r, topics); }).join("") + "</div></section>";
    }).join("");
    qs("[data-results]", results).innerHTML = html;
  }

  function fillEvidence() {
    qsa(".policy-cand", results).forEach(function (card) {
      var on = qs('.policy-topic[aria-pressed="true"]', card);
      var wrap = qs("[data-ev]", card);
      if (on && wrap) wrap.innerHTML = evidenceHTML(card.getAttribute("data-candidate"), on.getAttribute("data-topic"));
    });
  }

  function showResults(focus) {
    var ids = selectedIds();
    if (!ids.length) { show(total - 1, true); return; }
    writeHash(ids);
    renderResults(ids);
    quiz.hidden = true;
    results.hidden = false;
    root.classList.remove("pm-hash");
    if (!evidence && !evidenceFailed) evidenceReady.then(fillEvidence);
    if (focus) {
      window.scrollTo(0, 0);
      qs("#policy-results-title", results).focus({ preventScroll: true });
    }
  }

  function backToQuiz(clear) {
    if (clear) sel = {};
    writeHash([]);
    results.hidden = true;
    quiz.hidden = false;
    root.classList.remove("pm-hash");
    show(0, true);
    window.scrollTo(0, 0);
  }

  results.addEventListener("click", function (ev) {
    var topicBtn = ev.target.closest(".policy-topic");
    if (topicBtn) {
      var card = topicBtn.closest(".policy-cand");
      qsa(".policy-topic", card).forEach(function (b) { b.setAttribute("aria-pressed", b === topicBtn ? "true" : "false"); });
      qs("[data-ev]", card).innerHTML = evidenceHTML(card.getAttribute("data-candidate"), topicBtn.getAttribute("data-topic"));
      return;
    }
    var act = ev.target.closest("[data-act]");
    if (!act) return;
    var a = act.getAttribute("data-act");
    if (a === "share") MO.share(location.href, document.title);
    else if (a === "edit") backToQuiz(false);
    else if (a === "restart") backToQuiz(true);
  });

  function fromHash(focus) {
    var next = readHash();
    if (next && Object.keys(next).length) { sel = next; showResults(focus); return true; }
    return false;
  }
  window.addEventListener("hashchange", function () { fromHash(true); });

  if (!fromHash(false)) { root.classList.remove("pm-hash"); show(0, false); }
})();
